Translate Optimizing Performance (#110)
reactjs/pt-BR.reactjs.org: content/docs/optimizing-performance.md

---
id: optimizing-performance
title: Optimizando performance
permalink: docs/optimizing-performance.html
redirect_from:
  - "docs/advanced-performance.html"
---

Internamente, React usa diversas técnicas inteligentes para minimizar o número de operações custosas de DOM que são requiridas para alterar a UI. Para muitas aplicações que usam React, elas terão uma rápida interface sem fazer muito esforço para optimizar performance. No entanto, existem diversas maneiras para acelerar sua aplicação React.

## Usando em produção {#use-the-production-build}

Se vocuê está fazendo benchmarking ou tendo problemas de performance em suas aplicações React, tenha certeza que você está testando com a build de produção.

Por padrão, o React inclui diversos avisos úteis. Esses avisos são muito úteis em desenvolvimento. Contudo, eles tornam o React maior e mais lento, então você precisa ter certeza que está usando a versão de produção quando faz a publicação de seu app.

> Se você não tem certeza se seu processo de build está configurado corretamente, você pode checar instalando [a extensão React Developer Tools para o Chrome](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi). Se você visitar um site que usa React em produção, o ícone terá uma cor de fundo escura:

Reviewer: ```suggestion
Se você não tem certeza se seu processo de build está configurado corretamente, você pode verificar instalando a extensão [React Developer Tools para o Chrome](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi). Se você visitar um site que usa React em produção, o ícone terá uma cor de fundo escura:
```

```
# Se você usa npm
npm install --save-dev uglify-js-brunch

# Se você usar Yarn
yarn add --dev uglify-js-brunch
```

Então, para criar uma build de produção, adicione o argumento `-p` no comando `build`:

```
brunch build -p
```

Lembre que você somente precisa fazer isso para builds de produção. Você não deve passar o argumento `-p` ou aplicar esse plugin em desenvolvimento, porque ele irá esconder avisos úteis do React e fará as builds mais lentas.

### Browserify {#browserify}

Para uma Browserify build de produção mais eficiente, instale esses poucos plugins:

```
# Se você usa npm
npm install --save-dev envify uglify-js uglifyify 

# Se você usa Yarn
yarn add --dev envify uglify-js uglifyify 
```

Para criar uma build de produção, tenha certeza que você adicionou esses transforms, **(a ordem faz diferença)**

* O [`envify`](https://github.com/hughsk/envify) assegura que o ambiente que a build está configurado é o correto. Torne ele global (`-g`).
* O [`uglifyify`](https://github.com/hughsk/uglifyify) remove os imports de desenvolvimento. Torna ele global também (`-g`).

> * Finalmente, o bundle gerado é enviado para o [`uglify-js`](https://github.com/mishoo/UglifyJS2) para enxutar ([read why](https://github.com/hughsk/uglifyify#motivationusage)).

Reviewer 2: ```suggestion
* Finalmente, o bundle gerado é enviado para o [`uglify-js`](https://github.com/mishoo/UglifyJS2) para enxutar ([saiba por quê](https://github.com/hughsk/uglifyify#motivationusage)).
```

  | uglifyjs --compress --mangle > ./bundle.js
```

>**Observação:**
>
>O nome do pacote é `uglify-js`, mas o binário gerado é chamado `uglifyjs`.<br>
>Isto não é um erro de digitação.

Lembre que você somente precisar fazer isso para builds de produção. Você não deve aplicar esses plugins em desenvolvimento porque eles vão esconder avisos úteis do React, e farão as builds mais lentas.

### Rollup {#rollup}

> Para uma Rollup build de produção mais eficiente, instale esses poucos plugins:

Reviewer 2: ```suggestion
Para uma build de produção do Rollup mais eficiente, instale alguns plugins:
```

]
```

Para um exemplo completo de setup [veja esse gist](https://gist.github.com/Rich-Harris/cb14f4bc0670c47d00d191565be36bf0).

Lembre que você somente precisa fazer isso para builds de produção. Você não deve aplicar o `uglify` ou o `replace` com o valor de `'production'`em desenvolvimento porque eles vão esconder avisos úteis do React, e farão as builds mais lentas.

### webpack {#webpack}

>**Observação:**
>
>Se você está usando Create React App, por favor siga [as instruções abaixo](#create-react-app).<br>

> >Esta seção é somento relevante se você configura o webpack diretamente.

Reviewer 2: ```suggestion
>Esta seção é somente relevante se você configura o webpack diretamente.
```

> Para uma mais eficiente webpack build de produção, tenha certeza que você incluiu esses plugins em sua configuração de produção:

Reviewer 2: ```suggestion
Para uma build de produção mais eficiente do webpack, tenha certeza que você incluiu esses plugins em sua configuração de produção:
```

new webpack.optimize.UglifyJsPlugin()
```

Você pode aprender mais sobre isso na [documentação do webpack](https://webpack.js.org/guides/production-build/).

Lembre que você somente precisa fazer isso para builds de produção. Você não deve aplicar `UglifyJsPlugin` ou `DefinePlugin` com o valor de `'production'` em desenvolvimento porque eles vão esconder avisos úteis do React, e farão as builds mais lentas.

## Analisando componentes com o Chrome Performance Tab {#profiling-components-with-the-chrome-performance-tab}

> Em modo de **desenvolvimento**, você pode visualizar como os componentes são montados (mount), alterados (update), and desmontados (unmount), usando as ferramentas de perfomance nos browsers suportados. Por exemplo:

Reviewer 2: ```suggestion
Em modo de **desenvolvimento**, você pode visualizar como os componentes são montados (_mount_), alterados (_update_), and desmontados (_unmount_), usando as ferramentas de performance nos browsers suportados. Por exemplo:
```

> <center><img src="../images/blog/react-perf-chrome-timeline.png" style="max-width:100%" alt="Componentes React na linha do tempo do Chrome" /></center>

Reviewer 2: ```suggestion
<center><img src="../images/blog/react-perf-chrome-timeline.png" style="max-width:100%" alt="Componentes do React na linha do tempo do Chrome" /></center>
```

Para fazer isso no Chrome:

> 1. Temporariamente **desabilite todas as extensões do Chrome, especialmente React DevTools**. Eles podem significativamente enviesar os resultados!

Reviewer 2: ```suggestion
1. Temporariamente **desabilite todas as extensões do Chrome, especialmente React DevTools**. Elas podem significativamente enviesar os resultados!
```

2. Tenha certeza que você está rodando sua aplicação no modo de desenvolvimento.

3. Abra o Chrome DevTools a aba **[Performance](https://developers.google.com/web/tools/chrome-devtools/evaluate-performance/timeline-tool)** a pressione **Record**.

4. Faça as ações que você quer analisar. Não grave mais que 20 segundos ou o Chrome pode travar.

5. Pare de gravar.

> 6. Eventos React serão agrupados sobre a label **User Timing**.

Reviewer 2: ```suggestion
6. Eventos do React serão agrupados sob a label **User Timing**.
```

Para mais detalhes do passo a passo, veja [esse artigo do Ben Schwarz](https://calibreapp.com/blog/2017-11-28-debugging-react/).

Perceba que **os números são relativos para que os componentes renderizem mais rápido em produção**. Ainda, isto deve ajudar você a perceber quando algo não relacionados da UI são alteradas, a quão profundo e frequente suas alterações de UI acontecem.

Atualmente Chrome, Edge e IE são os únicos browsers que suportam essa feature, mas nós usamos um padrão [User Timing API](https://developer.mozilla.org/en-US/docs/Web/API/User_Timing_API) então esperamos que mais navegadores deem suporte.

> ## Analizando componentes com o DevTools Profiler {#profiling-components-with-the-devtools-profiler}

Reviewer 2: ```suggestion
## Analisando componentes com o DevTools Profiler {#profiling-components-with-the-devtools-profiler}
```

`react-dom` 16.5+ e `react-native` 0.57+ fornecem melhorias nas capacidades de analise em modo de desenvolvimento com o React DevTools Profiler.

Uma visão geral do Profiler pode ser encontrada nesse artigo ["Introducing the React Profiler"](/blog/2018/09/10/introducing-the-react-profiler.html).
Um vídeo com o passo a passo do profiler também está [disponível no YouTube](https://www.youtube.com/watch?v=nySib7ipZdk).

> Se você ainda não tem o React DevTools instalado, você pode fazer isso encontrado ele aqui:

Reviewer 2: ```suggestion
Se você ainda não tem o React DevTools instalado, você pode encontrá-lo aqui:
```

- [Extensão para Chrome](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en)
- [Extensão para Firefox](https://addons.mozilla.org/en-GB/firefox/addon/react-devtools/)
- [Pacote separado](https://www.npmjs.com/package/react-devtools)

> Observação
>

> > Um analise de uma build de produção do `react-dom` está disponível como `react-dom/profiling`.

Reviewer 2: ```suggestion
> Uma analise de uma build de produção do `react-dom` está disponível como `react-dom/profiling`.
```

> > Leia mais sobre como usar essa builda no [fb.me/react-profiling](https://fb.me/react-profiling)

Reviewer 2: ```suggestion
> Leia mais sobre como usar esse pacote no [fb.me/react-profiling](https://fb.me/react-profiling)
```

> ## Virtualizando longas listas {#virtualize-long-lists}

Reviewer 2: ```suggestion
## Virtualizando Longas Listas {#virtualize-long-lists}
```

> Se sua aplicação renderiza longas listas de informação (milhares ou centenas de linhas), nós recomendamos usar uma técnica conhecida como "windowing". Esta técnica somente renderiza um pequeno conjuto de suas linhas e pode drasticamente reduzir o tempo que ele leva para re-renderizar os compomentes bem como o número de nós DOM criados.

Reviewer 2: ```suggestion
Se sua aplicação renderiza longas listas de informação (milhares ou centenas de linhas), nós recomendamos usar uma técnica conhecida como "windowing". Esta técnica somente renderiza um pequeno conjunto de suas linhas e pode reduzir drasticamente o tempo que ele leva para re-renderizar os componentes bem como o número de nós criados no DOM.
```

[react-window](https://react-window.now.sh/) e [react-virtualized](https://bvaughn.github.io/react-virtualized/) são as bibliotecas de windowing mais populares. Eles fornecem diversos componentes reutilizáveis para exibir listas, grids e informações tabulares. Você pode também pode criar seu próprio componente de windowing, como [o Twitter fez](https://medium.com/@paularmstrong/twitter-lite-and-high-performance-react-progressive-web-apps-at-scale-d28a00e780a3), se você quer algo mais específico para sua aplicacão.

## Evite recompilação {#avoid-reconciliation}

> React cria e mantem sua representação interna da renderização UI. Ele inclui os elementos React dos seus componentes. Essa representação evita que o React crie nós DOM e acesse os existes sem necessidade, além do que essas operações podem ser mais lentas do que operações em objetos JavaScript. Algumas vezes esse processo é referenciado como "virtual DOM", mas ele funciona da mesma forma no React Native.

Reviewer 2: ```suggestion
O React cria e mantem sua representação interna da renderização UI. Ele inclui os elementos do React retornados pelos seus componentes. Essa representação evita que o React crie nós DOM e acesse os existes sem necessidade, além do que essas operações podem ser mais lentas do que operações em objetos JavaScript. Algumas vezes esse processo é referenciado como "virtual DOM", mas ele funciona da mesma forma no React Native.
```

> React cria e mantem sua representação interna da renderização UI. Ele inclui os elementos React dos seus componentes. Essa representação evita que o React crie nós DOM e acesse os existes sem necessidade, além do que essas operações podem ser mais lentas do que operações em objetos JavaScript. Algumas vezes esse processo é referenciado como "virtual DOM", mas ele funciona da mesma forma no React Native.

Reviewer 3: ```suggestion
O React cria e mantém sua representação interna da UI renderizada. Ele inclui os elementos do React que você retorna dos seus componentes. Essa representação evita que o React crie nós no DOM e acesse os existes sem necessidade, além do que essas operações podem ser mais lentas do que operações em objetos JavaScript. Algumas vezes esse processo é referenciado como "virtual DOM", mas ele funciona da mesma forma no React Native.
```

> Quando uma propriedade ou estado de um componente é alterado, o React decide se o DOM atual é alterado comparando o novo elemento retornado com o antigo. Quando eles não forem iguais, o React irá alterar o DOM.

Reviewer 2: ```suggestion
Quando uma propriedade ou estado de um componente é alterado, o React decide se uma atualização do DOM atual é necessária comparando o novo elemento retornado com o antigo. Quando eles não forem iguais, o React irá alterar o DOM.
```

Você pode agora visualizar essas re-renderizações do virtual DOM como o React DevTools:

- [Extensão para Chrome](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en)
- [Extensão para Firefox](https://addons.mozilla.org/en-GB/firefox/addon/react-devtools/)
- [Pacote separado](https://www.npmjs.com/package/react-devtools)

No console de desenvolvedor selecione a opção **Highlight Updates** na aba de **React**:

> <center><img src="../images/blog/devtools-highlight-updates.png" style="max-width:100%; margin-top:10px;" alt="Como habilitar os detaques de alteração" /></center>

Reviewer 2: ```suggestion
<center><img src="../images/blog/devtools-highlight-updates.png" style="max-width:100%; margin-top:10px;" alt="Como habilitar os destaques de alteração (highlight updates)" /></center>
```

> Interaja com sua página e você deve ver as bordas coloridas aparecendo envolta de qualquer componente que foi re-renderizado. Isto faz com que perceba re-renders que não são necessários. Você pode aprender mais sobre essa funcionalidade do React DevTools nesse [post](https://blog.logrocket.com/make-react-fast-again-part-3-highlighting-component-updates-6119e45e6833) do [Ben Edelstein](https://blog.logrocket.com/@edelstein).

Reviewer 2: ```suggestion
Interaja com sua página e você deve ver as bordas coloridas aparecendo ao redor de qualquer componente que foi re-renderizado. Isto faz com que você perceba re-renders que não são necessários. Você pode aprender mais sobre essa funcionalidade do React DevTools nesse [post](https://blog.logrocket.com/make-react-fast-again-part-3-highlighting-component-updates-6119e45e6833) do [Ben Edelstein](https://blog.logrocket.com/@edelstein).
```

Considere esse exemplo:

> <center><img src="../images/blog/highlight-updates-example.gif" style="max-width:100%; margin-top:20px;" alt="Exemplo do React DevTools Highlights de alterações" /></center>

Reviewer 2: ```suggestion
<center><img src="../images/blog/highlight-updates-example.gif" style="max-width:100%; margin-top:20px;" alt="Exemplo dos destaques de alterações do React DevTools" /></center>
```

> Perceba que quando nós estamos acessando o segundo todo, o primeiro todo também pisca na tela a cada tecla digitada. Isto significa que ele está sendo re-renderizando pelo React junto com o input. Isso algumas vez é chamado render desperdiçado (wasted render). Nós sabemos que ele é desnecessário porque o conteúdo do primeiro todo não tem nenhuma mudança, mas o React não sabe sobre isso.

Reviewer 2: ```suggestion
Perceba que quando nós estamos acessando o segundo _todo_, o primeiro todo também pisca na tela a cada tecla digitada. Isto significa que ele está sendo re-renderizando pelo React junto com o input. Isso algumas vez é chamado render desperdiçado (wasted render). Nós sabemos que ele é desnecessário porque o conteúdo do primeiro todo não tem nenhuma mudança, mas o React não sabe sobre isso.
```

> Embora o React somente altere os nós de DOM alterados, o re-rendering ainda leva algum tempo. Em muitos casos isso não é um problema, mas se a lentidão é perceptível, você pode aumentar velocidade dele sobrescrevendo a função de lifecycle `shouldComponentUpdate`, na qual é chamada antes do processo de re-rendering começar. A implementação padrão dessa função retorna `true`, deixando React performar a alteração:

Reviewer 2: ```suggestion
Embora o React somente altere os nós de DOM alterados, o re-rendering ainda leva algum tempo. Em muitos casos isso não é um problema, mas se a lentidão é perceptível, você pode aumentar velocidade dele sobrescrevendo a função de lifecycle `shouldComponentUpdate`, na qual é chamada antes do processo de re-rendering começar. A implementação padrão dessa função retorna `true`, deixando o React performar a alteração:
```

```javascript
shouldComponentUpdate(nextProps, nextState) {
  return true;
}
```

> Se você sabe que em algumas situações seu componente não precisa ser alterado, você pode retornar `false` no `shouldComponentUpdate` ao invés, para pular o todo o processo de renderização, incluindo o chamado `render()` nesse componente abaixo:

Reviewer 2: ```suggestion
Se você sabe que em algumas situações seu componente não precisa ser alterado, você pode retornar `false` no `shouldComponentUpdate` ao invés, para pular o todo o processo de renderização, incluindo a chamada de `render()` nesse componente e seus filhos:
```

Na maioria dos casos, ao invés de escrever `shouldComponentUpdate()` na mão, você pode herdar do [`React.PureComponent`](/docs/react-api.html#reactpurecomponent). Ele equivale a implementação do `shouldComponentUpdate()` com uma comparação rasa entre as anteriores e novas propriedades e estados 

> ## shouldComponentUpdate em ação {#shouldcomponentupdate-in-action}

Reviewer 2: ```suggestion
## shouldComponentUpdate Em Ação {#shouldcomponentupdate-in-action}
```

Reviewer 4: ```suggestion
## shouldComponentUpdate em Ação {#shouldcomponentupdate-in-action}
```

Preposições não precisam de capitalização.

> Aqui é uma sub-árvore de componentes. Para cada uma, `SCU` define o que o `shouldComponentUpdate` retorna, e `vDOMEq` indica se os elementos React renderizados são equivalentes. Finalmente, o círculo de cores indica se o componente tinha de ser rencoliado ou não.

Reviewer 2: ```suggestion
Aqui é uma sub-árvore de componentes. Para cada uma, `SCU` define o que o `shouldComponentUpdate` retorna, e `vDOMEq` indica se os elementos renderizados pelo React são equivalentes. Finalmente, o círculo de cores indica se o componente tinha de ser reconciliado ou não.
```

<figure><img src="../images/docs/should-component-update.png" style="max-width:100%" /></figure>

Já que `shouldComponentUpdate` retornou `false` na sub-árvore iniciada no C2, React não tentou renderizar C2, e por consequência não invocou `shouldComponentUpdate` no C4 e C5.

> Para C1 e C3, `shouldComponentUpdate` retornou `true`, então o React teve que descer até as folhas para checa-los. Para o C6 `shouldComponentUpdate` returned `true`, e já que os elementes renderizados não são iguais o React teve que alterar o DOM.

Reviewer 2: ```suggestion
Para C1 e C3, `shouldComponentUpdate` retornou `true`, então o React teve que descer até as folhas para checá-los. Para o C6 `shouldComponentUpdate` retornou `true`, e já que os elementos renderizados não são iguais, o React teve que alterar o DOM.
```

O último caso interessante é o C8. React teve que renderizar este componente, mas já que os elementos que ele retornou eram iguais aos previamente renderizados, ele não teve que alterar o DOM. 

> Note que o React somente tinha de fazer mutações no DOM para o C6, no qual era inevitável. Para C8, ele abortou comparando os elementos React renderizados, a para sub-árvore do C2 e C7, ele nem mesmo teve que comparar os elementos pois abordou no `shouldComponentUpdate`, e `render` não foi chamado.

Reviewer 2: ```suggestion
Note que o React somente teve de fazer mutações no DOM para o C6, no qual era inevitável. Para C8, ele abortou comparando os elementos React renderizados, e para a sub-árvore do C2 e C7, ele nem mesmo teve que comparar os elementos pois abortou no `shouldComponentUpdate`, e `render` não foi chamado.
```

}
```

> Nesse código, `shouldComponentUpdate` é somente checado se existe qualquer mudança no `props.color` ou `state.count`. Se esses valores não são alterados, o componente não é alterado. Se seu compoennte ficou complexo, você pode usar um padrão similar fazendo uma comparação rasa (shallow comparison) entre todos os fields de `props` e `state` para determinar se o componente deveria ser atualizado. Esse padrão é suficiente comum para que o React forneça um helper para usar essa lógica, apenas herdando de `React.PureComponent`. Então esse é uma simples maneira para alcançar isso:

Reviewer 2: ```suggestion
Nesse código, `shouldComponentUpdate` só está checando se houve alguma mudança no `props.color` ou `state.count`. Se esses valores não são alterados, o componente não é alterado. Se seu componente ficou complexo, você pode usar um padrão similar fazendo uma comparação rasa (shallow comparison) entre todos os fields de `props` e `state` para determinar se o componente deve ser atualizado. Esse padrão é comum o suficiente para que o React forneça um helper para usar essa lógica, apenas herdando de `React.PureComponent`. Então essa é uma maneira simples para alcançar isso:
```

Reviewer 4: ```suggestion
Nesse código, `shouldComponentUpdate` só está checando se houve alguma mudança no `props.color` ou `state.count`. Se esses valores não são alterados, o componente não é alterado. Se seu componente ficou mais complexo, você pode usar um padrão similar fazendo uma comparação rasa (shallow comparison) entre todos os fields de `props` e `state` para determinar se o componente deve ser atualizado. Esse padrão é comum o suficiente para que o React forneça um helper para usar essa lógica - apenas herde de `React.PureComponent`. Então, essa é uma maneira mais simples de alcançar a mesma coisa:```

Ajustando algumas semânticas e os superlativos.

}
```

> Na maioria do tempo, você pode usar `React.PureComponent` ao invés de escrever seu próprio `shouldComponentUpdate`. Ele somente faz comparações rasas, você não pode usar suas propriedades ou estado casos elas tenham sido alteradas de uma  maneira que a comparação rasa não irá detectar.

Reviewer 2: ```suggestion
Na maioria do tempo, você pode usar `React.PureComponent` ao invés de escrever seu próprio `shouldComponentUpdate`. Ele somente faz comparações rasas, então, você não pode usar suas propriedades ou estado casos elas tenham sido alteradas de uma maneira que a comparação rasa não irá detectar.
```

Reviewer 4: ```suggestion
Na maior parte das vezes, você pode usar `React.PureComponent` em vez de escrever seu próprio `shouldComponentUpdate`. Ele somente faz comparações rasas, então você não pode usá-lo caso as props ou state tenham sido alteradas de uma  maneira que a comparação rasa não iria detectar.
```

Ajustando tempos verbais.

**Obs:** "props" e "state" devem preferencialmente se manter sem tradução quando estão sendo referenciados diretamente, ok?

}
```

> O problema é que `PureComponent` vai fazer um comparação simples entre o valores antigos e novos de `this.props.words`. Já que esse código muta o lista de `words` no método `handleClick` do `WordAdder`, os antigos e novos valores de `this.props.words` irão ser compararados como iguais, mesmo que as atuais palavras da lista mudem. A `ListOfWords` não irá alterar mesmo que as novas palavras deveriam ser atualizadas.

Reviewer 2: ```suggestion
O problema é que `PureComponent` vai fazer um comparação simples entre o valores antigos e novos de `this.props.words`. Já que esse código muta a lista de `words` no método `handleClick` do `WordAdder`, os antigos e novos valores de `this.props.words` irão ser comparados como iguais, mesmo que as atuais palavras da lista mudem. A `ListOfWords` não irá alterar mesmo que as novas palavras deveriam ser atualizadas.
```

Reviewer 4: ```suggestion
O problema é que `PureComponent` vai fazer um comparação simples entre o valores antigos e novos de `this.props.words`. Já que esse código muta a lista de `words` no método `handleClick` do `WordAdder`, os antigos e novos valores de `this.props.words` serão comparados como iguais, mesmo que as atuais palavras da lista tenham mudado. A `ListOfWords` não irá alterar ainda que haja novas palavras que deveriam ser renderizadas.
```

Alguns ajustes nos tempos verbais.

> ## O poder não não mudar a informação {#the-power-of-not-mutating-data}

Reviewer 2: ```suggestion
## O Poder De Não Mutar Informação {#the-power-of-not-mutating-data}
```

Reviewer 4: ```suggestion
## O Poder de Não Mutar Dados {#the-power-of-not-mutating-data}
```
Em geral, as preposição não precisam ser capitalizadas nos títulos. E "data" está no plural então "dados" fica mais preciso. 👍

> A maneira mais simples desse problema não acontecer é evitar mutar valores que são usados como propriedades ou estado. Por exemplo, o método `handleClick` abaixo poderia ser rescrito usando `concact` como:

Reviewer 2: ```suggestion
A maneira mais simples desse problema não acontecer é evitar mutar valores que são usados como propriedades ou estado. Por exemplo, o método `handleClick` abaixo poderia ser reescrito usando `concat` como:
```

};
```

Você pode também reescrever o código que muta os objetos para evitar mutação, em uma maneira similar. Por exemplo, vamos dizer que nós temos um objeto chamado `colormap` e nós queremos escrever uma função que muda `colormap.right` para `'blue'`. Você poderia escrever:

```js
function updateColorMap(colormap) {
  colormap.right = 'blue';
}
```

Para escrever isso sem mutar o objeto original, nós poderíamos usar o método [Object.assign](https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/Global_Objects/Object/assign):

```js
function updateColorMap(colormap) {
  return Object.assign({}, colormap, {right: 'blue'});
}
```

`updateColorMap` agora retorna um novo objeto, ao invés de mutar o valor o antigo. `Object.assign` é ES6 e requer um polyfill.

Há uma proposta JavaScript para adicionar [espalhador de propriedades de objeto](https://github.com/sebmarkbage/ecmascript-rest-spread) para fazer ele alterar sem mutar.

```js
function updateColorMap(colormap) {
  return {...colormap, right: 'blue'};
}
```

> Se você está usando Create React App, ambos `Object.assign` e a sintaxe de espalhador de objeto estão disponível por padrão.

Reviewer 2: ```suggestion
Se você está usando Create React App, ambos `Object.assign` e a sintaxe de espalhador de objeto estão disponíveis por padrão.
```

> ## Usando estruturas da dados mutáveis {#using-immutable-data-structures}

Reviewer 2: ```suggestion
## Usando Estruturas De Dados Mutáveis {#using-immutable-data-structures}
```

x === y; // true
```

> Embora `y` seja editado, desde que sua referência para o objeto `x`, essa comparação retorna `true`. Você pode escrever um código similar com immutable.js

Reviewer 2: ```suggestion
Embora `y` foi editado, desde que sua referência para o objeto `x`, essa comparação retorna `true`. Você pode escrever um código similar com immutable.js:
```

x === z; // true
```

Nesse caso, já que uma nova referência é retornada quando mutamos `x`, nós podemos usar a referência para checar a equalidade `(x === y)` para verificar que o novo valor armazenado em `y` é diferente que o valor original em `x`.

Duas outras bibliotecas que pode ajudar o uso de dados imutáveis são [seamless-immutable](https://github.com/rtfeldman/seamless-immutable) e [immutability-helper](https://github.com/kolodny/immutability-helper).

> Estruturas de dados imutáveis fornecem para você uma maneira barata para rastrear mudanças em objetos, no qual é tudo que nós precisamos para implementar `shouldComponentUpdate`. Esse pode oferecer a você um bom impulsionamento de performance.

Reviewer 2: ```suggestion
Estruturas de dados imutáveis fornecem para você uma maneira barata para rastrear mudanças em objetos, no qual é tudo que nós precisamos para implementar `shouldComponentUpdate`. Isso pode oferecer a você um bom impulsionamento de performance.
```
